wolfcrypt/src/port/maxim/maxq10xx.c

```diff
@@ -0,0 +1,3384 @@
+    sha256.maxq_ctx.soft_hash = 1;
+
+    while (i < len) {
+        chunk = blkSz;
+        if ((chunk + i) > len) {
+            chunk = len - i;
+        }
+        /* Perform chunked update */
+        ret = wc_Sha256Update(&sha256, (buf + i), chunk);
+        if (ret != 0) {
+            break;
+        }
+        i += chunk;
+    }
+
+    if (ret == 0) {
+        /* Get final digest result */
+        ret = wc_Sha256Final(&sha256, hash);
+    }
+    return ret;
+}
+
+static int crypto_ecc_sign(const uint8_t *key, uint32_t keySz,
+    const uint8_t *hash, uint32_t hashSz, uint8_t *sig, uint32_t* sigSz,
+    uint32_t curveSz, int curveId, WC_RNG* rng)
+{
+    int ret;
+    mp_int r, s;
+    ecc_key ecc;
+
+    /* validate arguments */
+    if (key == NULL || hash == NULL || sig == NULL || sigSz == NULL ||
+        curveSz == 0 || hashSz == 0 || keySz < curveSz ||
+        *sigSz < (curveSz * 2)) {
+        return BAD_FUNC_ARG;
+    }
+
+    /* Initialize signature result */
+    XMEMSET(sig, 0, curveSz * 2);
+
+    /* Setup the ECC key */
+    ret = wc_ecc_init(&ecc);
+    if (ret < 0) {
+        return ret;
+    }
+
+    ecc.maxq_ctx.hw_ecc = -1;
+
+    /* Setup the signature r/s variables */
+    ret = mp_init(&r);
+    if (ret != MP_OKAY) {
+        wc_ecc_free(&ecc);
+        return ret;
+    }
+
+    ret = mp_init(&s);
+    if (ret != MP_OKAY) {
+        mp_clear(&r);
+        wc_ecc_free(&ecc);
+        return ret;
+    }
+
+    /* Import private key "k" */
+    ret = wc_ecc_import_private_key_ex(key, keySz, /* private key "d" */
+                                       NULL, 0,    /* public (optional) */
+                                       &ecc, curveId);
+
+    if (ret == 0) {
+        ret = wc_ecc_sign_hash_ex(hash, hashSz, /* computed hash digest */
+                                  rng, &ecc,    /* random and key context */
+                                  &r, &s);
+    }
+
+    if (ret == 0) {
+        /* export r/s */
+        mp_to_unsigned_bin_len(&r, sig, curveSz);
+        mp_to_unsigned_bin_len(&s, sig + curveSz, curveSz);
+    }
+
+    mp_clear(&r);
+    mp_clear(&s);
+    wc_ecc_free(&ecc);
+    return ret;
+}
+
+#ifdef MAXQ_DEBUG
+void dbg_dumphex(const char *identifier, const uint8_t* pdata, uint32_t plen)
+{
+    uint32_t i;
+
+    printf("%s\n", identifier);
+
+    for (i = 0; i < plen; ++i) {
+        if ((i > 0) && !(i % 16)) {
+            printf("\n");
+        }
+        printf("%02X ", pdata[i]);
+    }
+
+    printf("\n");
+}
+#endif /* MAXQ_DEBUG */
+
+/*
+ * Personalized Utility Functions
+ */
+static void HOST_LoadDefaultImportKey(unsigned char* key, int* keylen,
+                                      int* curve, int* type)
+{
+    *curve  = MXQ_KEYPARAM_EC_P256R1;
+    *type   = MXQ_KEYTYPE_ECC;
+    *keylen = 32;
+
+    XMEMCPY(key, KeyPairImport, sizeof(KeyPairImport));
+}
+
+static int getSignAlgoFromCurve(int c)
+{
+    switch(c) {
+        case MXQ_KEYPARAM_EC_P256R1:  return ALGO_ECDSA_SHA_256;
```

Comment:
Are Brainpool curves always supported?
Would be better if function fails when we aren't supporting the Brainpool curves in wolfSSL.